### DISCUSSION TITLE
How to get sub-dependencies when install or CPack?

### DISCUSSION THREAD

**Discussion by traversebitree** [microsoft/vcpkg#31686] [Q&A] [answered]
How to get sub-dependencies when **install** or **CPack**?
I use _vcpkg_ to manage the dependencies of my project. The `.dll` (dynamic link libraries) dependencies and sub-dependencies will be copied to my _build_ directory automatically. For example, my project depends on `libtins`, and `libtins` depends on `pcap.dll` and `libcrypto-3-x64.dll`. _vcpkg_ will copy `tins.dll`, `pcap.dll`, `libcrypto-3-x64.dll` to my _build_ directory automatically.

Now, I want to **install** and **CPack** my project, I use the following codes to install dependent `*.dll` to my _install_ directory.
```cmake
install(TARGETS tgt
    RUNTIME LIBRARY ARCHIVE COMPONENT "app_exe"
    DESTINATION ${CMAKE_INSTALL_BINDIR}
)
install(FILES $<TARGET_RUNTIME_DLLS:tgt>
    COMPONENT "app_dll"
    DESTINATION ${CMAKE_INSTALL_BINDIR}
)
```
But, only direct dependency `tins.dll` is being copied to my _install_ directory, sub-dependencies `pcap.dll`, `libcrypto-3-x64.dll` are not being copied to my _install_ directory.
I wonder to know how to deal with it.

![image](https://github.com/microsoft/vcpkg/assets/30373236/910295a2-6e96-4a40-8553-67f5a22f1e58)
![image](https://github.com/microsoft/vcpkg/assets/30373236/717273a8-1130-4a08-95fa-311ae5847d4b)

**Comment by traversebitree** [comment]
Similar question is here: [vcpkg: export all .dll dependencies + subdependencies of a package?](https://www.reddit.com/r/cpp_questions/comments/p9n1sb/vcpkg_export_all_dll_dependencies_subdependencies/)

**Reply by palonza** [reply]
Hi @traversebitree,

Here's how you can modify your CMakeLists.txt to include the vcpkg_export() function and handle sub-dependencies

```cmake
# Include vcpkg toolchain file
set(CMAKE_TOOLCHAIN_FILE "<path_to_vcpkg>/scripts/buildsystems/vcpkg.cmake" CACHE STRING "")

# Declare your project
project(YourProject)

# Find and include the necessary packages from vcpkg
find_package(tins CONFIG REQUIRED)

# Add your target
add_executable(YourTarget main.cpp)

# Link your target with the required libraries
target_link_libraries(YourTarget PRIVATE tins::tins)

# Export vcpkg environment variables and generate script
vcpkg_export(
  TARGETS ${_target_name}
  # Add any additional libraries or packages your project depends on here
)

# Install your target and its dependencies
install(TARGETS YourTarget
    RUNTIME LIBRARY ARCHIVE COMPONENT "app_exe"
    DESTINATION ${CMAKE_INSTALL_BINDIR}
)
install(FILES $<TARGET_RUNTIME_DLLS:YourTarget>
    COMPONENT "app_dll"
    DESTINATION ${CMAKE_INSTALL_BINDIR}
)
```

Replace <path_to_vcpkg> with the actual path to your vcpkg installation directory. This will ensure that the necessary sub-dependencies are copied to your install directory when using install() and CPack.

Make sure to run vcpkg integrate install in your project directory to integrate vcpkg with your CMake project before building and installing your project.

**Reply by traversebitree** [reply]
Thank you for your help, but the problem still remains. I've set `CMAKE_TOOLCHAIN_FILE`
![image](https://github.com/microsoft/vcpkg/assets/30373236/9bde80cb-5703-4427-8ea2-e6f71f2c3750)
But there's something wrong with the function or macro `vcpkg_export`.
![image](https://github.com/microsoft/vcpkg/assets/30373236/9ce1a1f4-d218-4639-b28a-32eba49bf8cb)

**Reply by traversebitree** [reply]
And I searched `vcpkg_export` in this repo, there's no this code.
![image](https://github.com/microsoft/vcpkg/assets/30373236/fd9bc375-e1d0-4c36-91ab-845de93f564b)

**Comment by palonza** [comment]
@traversebitree,

You can use (in your terminal) to export all dependencies and subdependencies:

```bash
vcpkg export --output=<output-directory>
```
in another way, you could use this in cmake (to execute any command from cmake):

```cmake
add_custom_command(
    TARGET <target>
    POST_BUILD
    COMMAND <cammand to execute at cmake> #<= vcpkg export ....
)
```
or you could use (cmake 3.2):

```cmake
execute_process()
```
After that, all exported files will be generated in the specified folder, then you could be use:

```cmake
find_package(<dependency-name> CONFIG REQUIRED)
```

make sure that run (to integrate dependencies and subdependencies):
```
vcpkg integrate install
```

**Comment by Osyotr** [comment]
Posts above look like they've been generated by some AI.
To install runtime dependencies you literally just need to add `RUNTIME_DEPENDENCIES` to your first `install` call and drop the second call.
Alternatively, use `RUNTIME_DEPENDENCY_SET deps` and then install it.

**Reply by traversebitree** [reply, marked as answer]
Thank you for your help. I tried this but the sub-dependencies are still not copied to the **install** directory.
I found the solution, which is an **experimental** option in _vcpkg_. `X_VCPKG_APPLOCAL_DEPS_INSTALL`
![image](https://github.com/microsoft/vcpkg/assets/30373236/270bbbdf-507e-4e13-bf61-3ffe69ec39f6)
By setting it to `ON` or `TRUE`, the sub-dependencies will be copied into the **install** directory.
I suggest the `X_VCPKG_APPLOCAL_DEPS_INSTALL` be set to `ON` by default.
![image](https://github.com/microsoft/vcpkg/assets/30373236/591945a5-26eb-48db-8413-6cd30dddcf5f)
